### PR TITLE
BoundDecisionDag.Rewrite - Avoid capturing the replacement map

### PR DESCRIPTION
**Pull request by alrz**
Extracted from https://github.com/dotnet/roslyn/pull/57909

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs [hunk 1]
```diff
@@ -623,7 +623,7 @@ void storeToTemp(BoundDagTemp temp, BoundExpression expr)
                     }
                 }
 
-                BoundDecisionDagNode makeReplacement(BoundDecisionDagNode node, Func<BoundDecisionDagNode, BoundDecisionDagNode> replacement)
+                static BoundDecisionDagNode makeReplacement(BoundDecisionDagNode node, IReadOnlyDictionary<BoundDecisionDagNode, BoundDecisionDagNode> replacement)
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs:626]: Could be inlined as lambda for caching (this is the tuple input optimization which is common compared to rewrite for constant input)

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs:626]: IMO when it comes to whether to use a lambda or to convert a local function to delegate, let's use whichever form we prefer to read, with the assumption that eventually the method group conversion to delegate caching will come about.

**Review by RikkiGibson — APPROVED** [review]

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 1)

**Comment by AlekseyTs** [comment]
@alrz Thank you for the contribution.

**Pull request merged**